### DISCUSSION TITLE
auth to gitlab at non-root path

### DISCUSSION THREAD

**Discussion by hisnamewasjaan** [profclems/glab#662] [Q&A]
When using `glab auth login`, I  am prompted for a hostname. But how do I specify an additional path? 
My customers gitlab is at `rep.company.dk/git`

**Comment by profclems** [comment]
Wow... I never expected an instance to have be hosted with a  `hostname/path` url. 

This will be creating a lot of issues when used with glab especially when fetching the project name from the git url. `glab` assumes that if the url is `git@gitlab.companyname.com:path1/path2.git` or `https://gitlab.companyname.com/path1/path2`, the project path is `path1/path2`.

Maybe we may have to consider this as an issue and see ways we can fix it. But it is advisable to use a subdomain for a GitLab self-hosted instance.

**Reply by hisnamewasjaan** [reply]
I see, thanks :) 
That is what i feared. 
Just to be sure, a typical url in our setup is still just like: `git@rep.companyname.dk:group/project.git`. 
That is, `/git` is not included.

**Reply by profclems** [reply]
Does the API endpoint use `rep.companyname.dk/api/v4` or `rep.companyname.dk/git/api/v4`?

**Reply by profclems** [reply]
If the api endpoint is `rep.companyname.dk/api/v4` (without `/git`), then you shouldn't have any problems using `glab`

**Reply by hisnamewasjaan** [reply]
The api endpoint is the same. Including `/git`. 
So it seems I am out of luck for now. I don't believe I'll be able to convince sys-admins to change the setup for this reason alone. As I understand it `rep.companyname.dk` is a proxy in front of gitlab and other stuff...

**Reply by profclems** [reply]
You're not out of lock... Just realized there's a way around it. Since the API host will need `/git` attached, you can't authenticate via `auth login` command but you can directly add the `token` and `api_host` to the config file or use the `config` command:

1. First add your auth token:
```sh
glab config set token <token-here> --host rep.companyname.dk
```
Note that the host should be without `/git`.

2. Set the `api_host`
```sh
glab config set api_host  rep.companyname.dk/git --host rep.companyname.dk
```
Note that the api_host has `/git`

**Reply by profclems** [reply]
Let me know if this helps

**Reply by profclems** [reply]
@hisnamewasjaan have you tried it?

**Reply by hisnamewasjaan** [reply]
I have now - doesn't seem to work right away though...
`glab mr list` gives me a 404 error: `GET https://rep.companyname.dk/git/api/v4/projects/git/group/myproject/merge_requests: 404 {message: 404 Project Not Found}`

**Reply by profclems** [reply]
I thought you said the git protocol is `git@rep.companyname.dk:group/project.git` instead of `git@rep.companyname.dk:git/group/project.git`? 

Are you using https or ssh? 

The project can't be found because `glab` using `git/group/myproject` instead of `group/myproject`.
I'm suspecting that you're using https and hence, the git url is probably `https://gitlab.companyname.com/git/group/project.git`. If that's the case, then you should use the ssh URL

**Reply by hisnamewasjaan** [reply]
hmm, well `git remote -v` reveals `git@rep.companyname.dk:group/project.git` as stated earlier. 
So ssh, right?

**Reply by hisnamewasjaan** [reply]
@profclems , any further insights on this? Or what I am missing?

**Reply by zemzale** [reply]
Can you maybe share your glab config, of course by removing all the sensitive stuff. Maybe it just fails because the config somehow is wrong.

**Reply by hisnamewasjaan** [reply]
Naturally... 
```
$ cat config.yml 
# What protocol to use when performing git operations. Supported values: ssh, https
git_protocol: ssh
# What editor glab should run when creating issues, merge requests, etc.  This is a global config that cannot be overridden by hostname.
editor:
# What browser glab should run when opening links. This is a global config that cannot be overridden by hostname.
browser:
# Set your desired markdown renderer style. Available options are [dark, light, notty] or set a custom style. Refer to https://github.com/charmbracelet/glamour#styles
glamour_style: dark
# Allow glab to automatically check for updates and notify you when there are new updates
check_update: false
# configuration specific for gitlab instances
hosts:
    gitlab.com:
        # What protocol to use to access the api endpoint. Supported values: http, https
        api_protocol: https
        # Configure host for api endpoint, defaults to the host itself
        api_host: gitlab.com
        # Your GitLab access token. Get an access token at https://gitlab.com/profile/personal_access_tokens
        token:
    rep.company.dk:
        git_protocol: ssh
        token: 8YgXXXXXXXYBHXnxrs
        api_host: rep.company.dk/git
```

Thank you for taking a look :)

**Reply by zemzale** [reply]
Well the config looks fine. Nothing pointing that something might be wrong. 

I am actually real puzzled, why it doesn't work. 

Could you try then a simple curl request and see if that works? Maybe there is some other weird reason why it doesn't work.
```sh
curl --request GET 'https://rep.company.dk/git/api/v4/projects/<id-of-project>/merge_requests' \
     --header 'Authorization: Bearer <your-token-here>'
```
